WIP: RP2040 PIO support

andryblack opened the pull request:
Initial support of RP2040 PIO hardware.
Compile time C++ assembler with syntax closed to original.

todo:
- [x] write all opcodes arguments
- [ ] work with dma

andryblack committed:
[rp] fix gpio functional

andryblack committed:
[rp] PIO with compile-time assembler

andryblack committed:
[rp] put PIO to GPIO at hal matrix

salkinium commented:
Very, very awesome btw! I also looked into this and wanted to be able to run the existing pio programs, so I wanted to create a pioasm template for modm, which I started work on here: https://github.com/modm-ext/pico-sdk-partial/tree/pioasm. 
I never finished because I realized that I needed the ASM part too, which you're working on now. Perhaps you can continue that if you want.

andryblack committed:
[rp] pio asm: simplify syntax, check encoding all instructions

andryblack commented:
> Very, very awesome btw! I also looked into this and wanted to be able to run the existing pio programs, so I wanted to create a pioasm template for modm, which I started work on here: https://github.com/modm-ext/pico-sdk-partial/tree/pioasm. I never finished because I realized that I needed the ASM part too, which you're working on now. Perhaps you can continue that if you want.

The main idea is not using tool for generating PIO code. Existing PIO programs need adapted in any case, because constaint framework specific initialisation code.  Programs code not big in size and ported as one to one. I updated syntax: removed excess braces.

andryblack committed:
[rp] pio: make initialization code more close to pio program

andryblack committed:
[rp] pio: fix whitespaces

andryblack committed:
[rp] pio: fix formatting
